api/comms/websocket.go

```diff
@@ -0,0 +1,268 @@
+	set := m.clients[cl.userId]
+	if set != nil {
+		if _, ok := set[cl]; ok {
+			delete(set, cl)
+			if len(set) == 0 {
+				delete(m.clients, cl.userId)
+			}
+		}
+	}
+	// Close underlying connection and channels
+	_ = cl.conn.Close()
+	select {
+	case <-cl.quit:
+		// already closed
+	default:
+		close(cl.quit)
+	}
+}
+
+func (cl *Client) readPump() {
+	// Keep the connection alive by consuming control/data frames and handling pongs.
+	cl.conn.SetReadLimit(maxIncomingMsgSize)
+	_ = cl.conn.SetReadDeadline(time.Now().Add(readIdleTimeout))
+	cl.conn.SetPongHandler(func(string) error {
+		return cl.conn.SetReadDeadline(time.Now().Add(readIdleTimeout))
+	})
+
+	// We don't expect app-level inbound messages, but we still need to read to
+	// receive pings/closes and keep deadlines fresh.
+	for {
+		mt, r, err := cl.conn.NextReader()
+		if err != nil {
+			cl.manager.logger.Debug("ws read closed",
+				zap.Int32("userId", cl.userId),
+				zap.Error(err))
+			cl.manager.removeClient(cl)
+			return
+		}
+		// Discard payload if any; we only care about keeping the socket healthy.
+		switch mt {
+		case websocket.TextMessage, websocket.BinaryMessage:
+			// Efficiently drain without allocating:
+			buf := make([]byte, 1024)
+			for {
+				_, derr := r.Read(buf)
+				if derr != nil {
+					break
+				}
+			}
+		}
+	}
+}
+
+func (cl *Client) writePump() {
+	ticker := time.NewTicker(pingInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case msg, ok := <-cl.send:
+			if !ok {
+				// Manager closed channel; send a close frame if possible.
+				_ = cl.conn.SetWriteDeadline(time.Now().Add(writeDeadline))
+				_ = cl.conn.WriteMessage(websocket.CloseMessage, nil)
+				cl.manager.removeClient(cl)
+				return
+			}
+			_ = cl.conn.SetWriteDeadline(time.Now().Add(writeDeadline))
+			if err := cl.conn.WriteMessage(websocket.TextMessage, msg); err != nil {
+				cl.manager.logger.Info("ws write error",
+					zap.Int32("userId", cl.userId),
+					zap.Error(err))
+				cl.manager.removeClient(cl)
+				return
+			}
+
+		case <-ticker.C:
+			// Keep-alive ping
+			_ = cl.conn.SetWriteDeadline(time.Now().Add(writeDeadline))
+			if err := cl.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
+				cl.manager.logger.Debug("ws ping failed",
+					zap.Int32("userId", cl.userId),
+					zap.Error(err))
+				cl.manager.removeClient(cl)
+				return
+			}
+
+		case <-cl.quit:
+			// Manager requested shutdown
+			return
+		}
+	}
+}
+
+// Push to a single receiver (all connected clients)
+func (m *CommsWebsocketManager) WebsocketPush(senderUserId int32, receiverUserId int32, rpcJson json.RawMessage, timestamp time.Time) {
+	encodedSenderUserId, _ := trashid.EncodeHashId(int(senderUserId))
+	encodedReceiverUserId, _ := trashid.EncodeHashId(int(receiverUserId))
+
+	data := struct {
+		RPC      json.RawMessage `json:"rpc"`
+		Metadata Metadata        `json:"metadata"`
+	}{
+		rpcJson,
+		Metadata{
+			Timestamp:      timestamp.Format(time.RFC3339Nano),
```

Comment:
what's this format?

Reply:
Taken from existing implementation: https://github.com/AudiusProject/audius-protocol/blob/49d271631bcc6b3528a48d6d377cbfb6e53c6459/comms/discovery/rpcz/websocket.go#L77

It maps to an ISO timestamp with nanosecond accuracy (🙄 ). I am not sure if we break clients by changing it.